pkg/scheduler/core/generic_scheduler.go

```diff
@@ -984,6 +983,10 @@ func selectVictimsOnNode(
 	queue internalqueue.SchedulingQueue,
 	pdbs []*policy.PodDisruptionBudget,
 ) ([]*v1.Pod, int, bool) {
+	if nodeInfo.Node() == nil {
```

Comment:
Which line of Clone do we think is causing the panic here?  Looking at the nodeInfo.Clone() method body, I couldn't immediately spot which copy would panic on a stripped node, unless `nodeInfo` itself was nil.

Reply:
@justinsb from the stack trace, it's:

https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/pkg/scheduler/cache/node_info.go#L437

but honestly I can't find any code related with "setting requestedResource to nil" or "build NodeInfo() without init requestedResource".

And from the trace, it seems `nodeInfo` itself is not nil.

Reply:
@Huang-Wei I think the fix is to change `NodeInfo.Clone()` code to check for nil before calling Clone() on `requestedResource`, `nonzeroRequest`, and `allocatableResource`.

Reply:
Yes, we can do that - that's a more safe way.

My point on checking _if NodeInfo.node is nil_ was: if "nodeInfo.node == nil" is related with "nodeInfo.requestedResource == nil", (we haven't been able to confirm it yet), then we can directly return as calculation on a NodeInfoCopy with node==nil sounds error prone, or maybe we will see other panics.

As of now, I think we can make both changes on (1) checking "if NodeInfo.node is nil" and (2) if "requestedResource/nonzeroRequest/allocatableResource" is nil? Thoughts?

Reply:
I have no reason to believe that checking NodeInfo.node is nil helps with mitigating the crash. I would limit this to checking resources. Alternatively, we could dig to find where we initialize a NodeInfo object without initialing the "resources"

Reply:
ok, I will update to only check resources.

Reply:
Hmm, one interesting thing after reading the trace carefully, sharing with you: https://github.com/kubernetes/kubernetes/issues/70450#issuecomment-439253820

Reply:
Yes, I had seen that in the trace and that's why I suggested the fix above.

Reply:
> Based on above comment, and that there are no chances requestedResource can be nil, I prefer to think the panic comes from a nil *NodeInfo.

I agree, I think it's the nodeInfo being nil causing the issue.

There are couple of ways to solve this, either at call site(place we are calling `nodeInfo.Clone()`, which are doing now in this PR) or within every method of nodeInfo where we can check if nodeInfo is nil. I don't have strong opinion as I see there are advantages of failing early(at the call site, which helps in friendly stack trace while debugging) going to method and checking it there(there is a good chance that someone might miss guarding against nil pointer when they are interacting with existing method assuming it is not going to panic, which happened in this case). 

If you want to do it at call site(like you are doing here), I'd recommend adding check for `nodeInfo==nil || nodeInfo.Node() == nil`

Reply:
> I'd recommend adding check for `nodeInfo==nil || nodeInfo.Node() == nil`

nodeInfo.Node() checks both if "nodeInfo == nil" or "nodeInfo.node == nil" :)

Reply:
How does it check both? Besides, I am not sure if we should actually check nodeInfo.node == nil.

Reply:
```go
func (n *NodeInfo) Node() *v1.Node {
	if n == nil {
		return nil
	}
	return n.node
}
```

nodeInfo.node == nil may not happen, as nodeInfo here is a cloned one, setting nodeInfo.node to nil can only happen in schedulerCache when it receive event of RemoveNode(). That one won't impact this cloned one.

So @bsalamat @ravisantoshgudimetla I can change it to check if nodeInfo == nil only, if it's more reasonable to you.

Reply:
I see your point Wei. You are right, but I still feel checking nodeInfo == nil is more readable for those who may be confused like myself.

Reply:
sure, will update.

Reply:
Yeah, at first I found it be confusing too but after looking at the code for a bit, I understood it.